test/integration/client/exec_test.go

```diff
@@ -0,0 +1,329 @@
+		outputEnvVar = "EXEC_PLUGIN_OUTPUT"
+	)
+
+	const (
+		clientAuthorizedToken   = "authorized-token"
+		clientUnauthorizedToken = "unauthorized-token"
+	)
+
+	certDir, err := ioutil.TempDir("", "kubernetes-client-exec-test-cert-dir-*")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tokenFileName := writeTokenFile(t, clientAuthorizedToken)
+	clientCAFileName, clientSigningCert, clientSigningKey := writeCACertFiles(t, certDir)
+	clientCertFileName, clientKeyFileName := writeCerts(t, clientSigningCert, clientSigningKey, certDir, 30*time.Second)
+	result := kubeapiservertesting.StartTestServerOrDie(
+		t,
+		nil,
+		[]string{
+			"--token-auth-file", tokenFileName,
+			"--client-ca-file=" + clientCAFileName,
+		},
+		framework.SharedEtcd(),
+	)
+	defer result.TearDownFn()
+
+	unauthorizedCert, unauthorizedKey, err := cert.GenerateSelfSignedCertKey("some-host", nil, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name                  string
+		clientConfigFunc      func(*rest.Config)
+		wantClientErrorPrefix string
+	}{
+		{
+			name: "unauthorized token",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
+					Name: outputEnvVar,
+					Value: `{
+										"kind": "ExecCredential",
+										"apiVersion": "client.authentication.k8s.io/v1beta1",
+										"status": {
+											"token": "unauthorized"
+										}
+									}`,
+				})
+			},
+			wantClientErrorPrefix: "Unauthorized",
+		},
+		{
+			name: "unauthorized certificate",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
+					Name: outputEnvVar,
+					Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"clientCertificateData": %q,
+								"clientKeyData": %q
+							}
+						}`, unauthorizedCert, unauthorizedKey),
+				})
+			},
+			wantClientErrorPrefix: "Unauthorized",
+		},
+		{
+			name: "authorized token",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
+					Name: outputEnvVar,
+					Value: fmt.Sprintf(`{
+						"kind": "ExecCredential",
+						"apiVersion": "client.authentication.k8s.io/v1beta1",
+						"status": {
+							"token": "%s"
+						}
+					}`, clientAuthorizedToken),
+				})
+			},
+		},
+		{
+			name: "authorized certificate",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = append(c.ExecProvider.Env, clientcmdapi.ExecEnvVar{
+					Name: outputEnvVar,
+					Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"clientCertificateData": %s,
+								"clientKeyData": %s
+							}
+						}`, read(t, clientCertFileName), read(t, clientKeyFileName)),
+				})
+			},
+		},
+		{
+			name: "authorized token and certificate",
```

Comment:
mm should we check that this favors token somehow?

Reply:
i thought if the exec plugin returns both a token and a cert, then client-go will use both the cert (as a client cert) and the token (in an http header). so when you say "favors token" do you mean asserting that the identity extracted by the apiserver is from the token and not the cert?

Reply:
Right.  `client-go` should always use both when specified.  `favors` here refers to the API server itself.

Reply:
281f7f9f53ed88ce8ab564967d7deac4e25fcf4b

We had a side convo over Zoom, and decided it made sense to only assert behavior from the client in this integration test. Therefore, we added specific assertions about the token and cert used by the client when making requests.